### PR TITLE
Add knative/pkg based TektonPipeline API and Reconciler

### PR DESCRIPTION
**Pull request by nikhil-thomas**
This is a `XXL` patch (on `main branch`) as it,  

- bootstraps knative/pkg based operator.
- adds TektonPIpeline API
- adds TektonPipeline Reconciler (matches current master TektonPipeline Reconciler functionality)

I shall add the tests, workflow scripts and docs in a separate commit

Code Commits
https://github.com/tektoncd/operator/pull/143/commits/be9becc035f7d55646928f4eff2f28c32fbcd25d
https://github.com/tektoncd/operator/pull/143/commits/9580fb9c336647a6956e2e2c22edba2a33a5ccc4
https://github.com/tektoncd/operator/pull/143/commits/f00a581e01dc1a8ee7714e4f0b4291e843d9ee44
https://github.com/tektoncd/operator/pull/143/commits/9cae5d49a7ac5f4e6c436a91bc455cc45b38ab24
https://github.com/tektoncd/operator/pull/143/commits/b675a9db8d81519dacdab3c79e6bb6d42b81d328
https://github.com/tektoncd/operator/pull/143/commits/94e23d1cd284b2ebd5f8b2a76afa6302a49835a8
https://github.com/tektoncd/operator/pull/143/commits/4c204621059e958d555945b1646aec1d9ed32077

Vendor Commits
https://github.com/tektoncd/operator/pull/143/commits/3ca889ded966bc568741e268cea89357334a628d
https://github.com/tektoncd/operator/pull/143/commits/bd24a0b28354b60d1313617b9309b5b491db0644
https://github.com/tektoncd/operator/pull/143/commits/8f0b601953d11852a447e01d7bd96d05d2444f39

# Changes

Add TektonPipeline API
    
Add TektonPipeline API with spec for installing TektonPipeline

Add codegen scripts

Add generated clients and deepcopy

Add generated clientset, informers, injection clients, and listers

Add generated `deepcopy`

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
unit tests
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

### CODE REVIEW

**Review by vincent-pli — APPROVED** [review]
Great, thanks

**Review by savitaashture — COMMENTED** [review]
PR looks good to me 
Thank you :slightly_smiling_face: 

There are few observation
1. I see few testcases are commented
2. Update the `Copyright 2019` to `Copyright 2020` as there are few files which are added as part of this PR

**Comment by nikhil-thomas** [comment]
> There are few observation
1. I see few testcases are commented
2. Update the Copyright 2019 to Copyright 2020 as there are few files which are added as part of this PR

@savitaashture 

1. I shall work on the tests right after this patch. Most of the code in `pkg/reconciler/common/` came from Knative operator. The reason why the tests they are commented is because of a key difference between KnativeServing (namespaces) and TektonPipeline (clusterwide) CRDs. I shall fix the tests in the next pr.

2. Update the Copyright 2019 to Copyright 2020 as there are few files which are added as part of this PR
good catch :P 
i copied the copyright from already existing files in Tekton Operator (which are old)
let me fix it right away :)

**Review by vdemeester — COMMENTED** [review]
Few comments :wink: 
Nice work :clap:

File: go.mod [hunk 1]
```diff
@@ -1,5 +1,28 @@
...
 
 go 1.14
+
+replace k8s.io/api => k8s.io/api v0.17.6
+
+replace k8s.io/apimachinery => k8s.io/apimachinery v0.17.6
+
+replace k8s.io/client-go => k8s.io/client-go v0.17.6
+
+replace k8s.io/code-generator => k8s.io/code-generator v0.17.6
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply go.mod]: done

File: pkg/reconciler/common/deployments_test.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+//	fake "github.com/manifestival/manifestival/fake"
+//	appsv1 "k8s.io/api/apps/v1"
+//	corev1 "k8s.io/api/core/v1"
+//	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+//	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+//	"k8s.io/apimachinery/pkg/runtime"
+//	v1alpha1 "knative.dev/operator/pkg/apis/operator/v1alpha1"
+//)
+//
+//func TestCheckDeployments(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply pkg/reconciler/common/deployments_test.go]: removed

File: hack/update-codegen.sh [hunk 3]
```diff
@@ -18,32 +18,42 @@ set -o errexit
...
-# #                  instead of the $GOPATH directly. For normal projects this can be dropped.
-# ${CODEGEN_PKG}/generate-groups.sh "deepcopy,client,informer,lister" \
-#   github.com/tektoncd/pipeline/pkg/client github.com/tektoncd/pipeline/pkg/apis \
-#   pipeline:v1alpha1 \
-#   --go-header-file ${REPO_ROOT_DIR}/hack/boilerplate/boilerplate.go.txt
+# generate the code with:
+# --output-base    because this script should also be able to run inside the vendor dir of
+#                  k8s.io/kubernetes. The output-base is needed for the generators to output into the vendor dir
+#                  instead of the $GOPATH directly. For normal projects this can be dropped.
+chmod +x ${CODEGEN_PKG}/generate-groups.sh
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply hack/update-codegen.sh:40]: removed

File: hack/update-codegen.sh [hunk 4]
```diff
@@ -18,32 +18,42 @@ set -o errexit
...
+  github.com/tektoncd/operator/pkg/client github.com/tektoncd/operator/pkg/apis \
+  "operator:v1alpha1" \
+  --go-header-file ${REPO_ROOT_DIR}/hack/boilerplate/boilerplate.go.txt
 
-# # Depends on generate-groups.sh to install bin/deepcopy-gen
-# ${GOPATH}/bin/deepcopy-gen \
-#   -O zz_generated.deepcopy \
-#   --go-header-file ${REPO_ROOT_DIR}/hack/boilerplate/boilerplate.go.txt \
-#   -i github.com/tektoncd/pipeline/pkg/apis/config
+chmod +x ${KNATIVE_CODEGEN_PKG}/hack/generate-knative.sh
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply hack/update-codegen.sh:46]: removed

File: hack/update-deps.sh [hunk 5]
```diff
@@ -14,24 +14,25 @@
...
+source $(dirname $0)/../vendor/knative.dev/test-infra/scripts/library.sh
+
+cd "${REPO_ROOT_DIR}"
 
 # The list of dependencies that we track at HEAD and periodically
 # float forward in this repository.
 FLOATING_DEPS=(
+  "knative.dev/pkg@release-0.17"
+  "knative.dev/test-infra@release-0.17"
 )
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply hack/update-deps.sh:36]: done

File: hack/verify-codegen.sh [hunk 6]
```diff
@@ -18,7 +18,7 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-source $(dirname $0)/../vendor/github.com/tektoncd/plumbing/scripts/library.sh
+source $(dirname $0)/../vendor/knative.dev/test-infra/scripts/library.sh
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply hack/verify-codegen.sh:21]: ooops, got it right in `udate-codegen.sh` but missed it here
(fixed)

File: hack/tools.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+limitations under the License.
+*/
+
+package tools
+
+// This package imports things required by this repository, to force `go mod` to see them as dependencies
+import (
+	_ "k8s.io/code-generator"
+	_ "knative.dev/pkg/hack"
+	_ "knative.dev/test-infra/scripts"
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply hack/tools.go]: done

File: pkg/reconciler/common/install_test.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+//import (
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply pkg/reconciler/common/install_test.go]: done

File: pkg/reconciler/common/releases.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+	//manifests := instance.GetSpec().GetManifests()
+	//// Create the comma-separated string as the URL to retrieve the manifest
+	//urls := make([]string, 0, len(manifests))
+	//for _, manifest := range manifests {
+	//	url := strings.ReplaceAll(manifest.Url, VersionVariable, version)
+	//	urls = append(urls, url)
+	//}
+	//return strings.Join(urls, COMMA)
+	return ""
+}
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply pkg/reconciler/common/releases.go:186]: this is not necessary at present.
so removed.

File: pkg/reconciler/common/releases.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+
+func createManifestsPath(instance v1alpha1.TektonComponent) []string {
+	//if len(instance.GetSpec().GetManifests()) > 0 {
+	//	return strings.Split(manifestPath(TargetVersion(instance), instance), COMMA)
+	//}
+
+	return make([]string, 0, 0)
+}
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply pkg/reconciler/common/releases.go:194]: this is not necessary at present.
so removed.

File: pkg/reconciler/common/releases.go [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+	if current == "" {
+		return true
+	}
+	current = sanitizeSemver(current)
+	target := sanitizeSemver(TargetVersion(instance))
+
+	currentMajor := semver.Major(current)
+	targetMajor := semver.Major(target)
+	if currentMajor != targetMajor {
+		// All the official releases of Knative are under the same Major version number. If target and current versions
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply pkg/reconciler/common/releases.go:80]: this is not necessary at present.
so removed.

File: hack/update-deps.sh [hunk 12]
```diff
@@ -14,24 +14,25 @@
...
+
+source $(dirname $0)/../vendor/knative.dev/test-infra/scripts/library.sh
+
+cd "${REPO_ROOT_DIR}"
 
 # The list of dependencies that we track at HEAD and periodically
 # float forward in this repository.
 FLOATING_DEPS=(
+  "knative.dev/pkg@release-0.17"
+  "knative.dev/test-infra@release-0.17"
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply hack/update-deps.sh]: done

File: hack/update-codegen.sh [hunk 13]
```diff
@@ -18,32 +18,42 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+export GO111MODULE=on
+# If we run with -mod=vendor here, then generate-groups.sh looks for vendor files in the wrong place.
+export GOFLAGS=-mod=
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply hack/update-codegen.sh:23]: `export GOFLAGS=-mod=` this turned out to be unnecesary.
it works either way. I didn't try without the flag change

so removed it

**Comment by nikhil-thomas** [comment]
@vdemeester thanks for the comments :+1: 
i have addressed all your comments and pushed the updates. :technologist: :man_dancing:

**Review by vdemeester — COMMENTED** [review]
/lgtm
/meow
:tada:

**Pull request merged**